File: src/core/logic/sheet.py
"""Module for handling spreadsheet workbooks and sheets."""

from pathlib import Path
from openpyxl import load_workbook
from openpyxl.workbook import Workbook

from src.config import SPREADSHEET_EXTENSIONS
from src.core.exceptions import FileExtensionNotSupported, IndexOutOfRange


class SheetGroup:
    """Represents a workbook file."""

    def __init__(self, file_path: Path):
        ext = Path(file_path).suffix.lstrip('.')
        if ext not in SPREADSHEET_EXTENSIONS:
            raise FileExtensionNotSupported(ext)

        self._workbook = load_workbook(filename=file_path, read_only=True, data_only=True)
        self._tables = {sheet_name: SheetTable(self._workbook, sheet_name) for sheet_name in self._workbook.sheetnames}

    @property
    def workbook(self):
        """Get the workbook object."""
        return self._workbook

    @property
    def tables(self):
        """Get all tables in the workbook."""
        return self._tables

    @property
    def file_path(self):
        """Get the file path of the workbook."""
        return self._workbook.path

    @property
    def name(self):
        """Get the name of the workbook file."""
        return self._workbook.properties.title


class SheetTable:
    """Represents a table in a sheet (in workbook file)."""

    def __init__(self, workbook: Workbook, sheet_name: str):
        self._sheet = workbook[sheet_name]

        for row in self._sheet.iter_rows():
            for cell in row:
                if cell.value is not None:
                    if self._min_row is None or cell.row < self._min_row:
                        self._min_row = cell.row
                    if self._max_row is None or cell.row > self._max_row:
                        self._max_row = cell.row
                    if self._min_col is None or cell.column < self._min_col:
                        self._min_col = cell.column
                    if self._max_col is None or cell.column > self._max_col:
                        self._max_col = cell.column

        self._headers = [
            self._sheet.cell(row=self._min_row, column=col).value
            for col in range(self._min_col, self._max_col + 1)
        ]

    @property
    def sheet(self):
        """Get the sheet object."""
        return self._sheet

    @property
    def headers(self) -> list[str]:
        """Get the headers of the table."""
        return self._headers

    @property
    def amounts(self) -> int:
        """Get the number of data rows in the table."""
        return self._max_row - self._min_row

    @property
    def name(self) -> str:
        """Get the name of the sheet."""
        return self._sheet.title

    def get_row(self, index: int):
        """Get a row by index (1-based)."""
        if index < 1 or index > self.amounts:
            raise IndexOutOfRange(index, (1, self.amounts))

        index += self._min_row  # Adjust index to actual row number

        row_data: dict[str, str] = {}
        for col in range(self._min_col, self._max_col + 1):
            header = self.headers[col - self._min_col]
            cell_value = self._sheet.cell(row=index, column=col).value
            row_data[header] = cell_value
        return row_data

File: data/Models/Sheet.cs
using System.Collections.Concurrent;
using ClosedXML.Excel;
using TaoSlideTotNghiep.Config;
using TaoSlideTotNghiep.Exceptions;

namespace TaoSlideTotNghiep.Models;

/// <summary>
/// Represents a workbook file and its sheets.
/// </summary>
public class Workbook : IDisposable
{
    private readonly XLWorkbook _workbook;
    private readonly ConcurrentDictionary<string, Worksheet> _sheets;
    private bool _disposed;

    public string FilePath { get; }

    /// <summary>
    /// Initializes a Workbook by loading the workbook from the specified file path.
    /// </summary>
    public Workbook(string filePath)
    {
        var ext = Path.GetExtension(filePath).TrimStart('.').ToLowerInvariant();

        if (!AppConfig.SpreadsheetExtensions.Contains(ext))
            throw new FileExtensionNotSupportedException(ext);

        FilePath = filePath;
        _workbook = new XLWorkbook(filePath);
        _sheets = new ConcurrentDictionary<string, Worksheet>();

        // Load all sheets as tables
        foreach (var worksheet in _workbook.Worksheets)
            _sheets[worksheet.Name] = new Worksheet(worksheet);
    }

    /// <summary>
    /// Gets all tables in the workbook.
    /// </summary>
    public IReadOnlyDictionary<string, Worksheet> Sheets => _sheets;

    /// <summary>
    /// Gets the workbook name.
    /// </summary>
    public string? Name => _workbook.Properties.Title;

    /// <summary>
    /// Gets table names with their row counts.
    /// </summary>
    public Dictionary<string, int> GetTableInfo()
    {
        return _sheets.ToDictionary(t => t.Key, t => t.Value.RowCount);
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _workbook.Dispose();
        GC.SuppressFinalize(this);
    }
}

/// <summary>
/// Represents a worksheet (table) in a workbook.
/// </summary>
public class Worksheet
{
    private readonly IXLWorksheet _worksheet;
    private readonly int _minRow;
    private readonly int _maxRow;
    private readonly int _minCol;
    private readonly int _maxCol;
    private readonly List<string?> _headers;

    public string Name => _worksheet.Name;
    public IReadOnlyList<string?> Headers => _headers;
    public int RowCount => _maxRow - _minRow;

    public Worksheet(IXLWorksheet worksheet)
    {
        _worksheet = worksheet;

        // Find the used range bounds
        var usedRange = worksheet.RangeUsed();
        if (usedRange == null)
        {
            _minRow = _maxRow = _minCol = _maxCol = 1;
            _headers = [];
            return;
        }

        _minRow = usedRange.FirstRow().RowNumber();
        _maxRow = usedRange.LastRow().RowNumber();
        _minCol = usedRange.FirstColumn().ColumnNumber();
        _maxCol = usedRange.LastColumn().ColumnNumber();

        // Extract headers from first row
        _headers = [];
        for (var col = _minCol; col <= _maxCol; col++)
        {
            var cellValue = worksheet.Cell(_minRow, col).GetValue<string>();
            _headers.Add(cellValue);
        }
    }

    /// <summary>
    /// Gets a row by index (1-based, relative to data rows).
    /// </summary>
    public Dictionary<string, object?> GetRow(int index)
    {
        if (index < 1 || index > RowCount)
            throw new Exceptions.IndexOutOfRangeException(index, (1, RowCount));

        var actualRow = _minRow + index; // Skip header row
        var rowData = new Dictionary<string, object?>();

        for (var col = _minCol; col <= _maxCol; col++)
        {
            var header = _headers[col - _minCol];
            var cellValue = _worksheet.Cell(actualRow, col).Value;

            // Convert XLCellValue to appropriate type
            object? value = cellValue.Type switch
            {
                XLDataType.Text => cellValue.GetText(),
                XLDataType.Number => cellValue.GetNumber(),
                XLDataType.Boolean => cellValue.GetBoolean(),
                XLDataType.DateTime => cellValue.GetDateTime(),
                XLDataType.TimeSpan => cellValue.GetTimeSpan(),
                _ => cellValue.ToString()
            };

            rowData[header ?? $"Column{col}"] = value;
        }

        return rowData;
    }

    /// <summary>
    /// Gets all rows as a list of dictionaries.
    /// </summary>
    public List<Dictionary<string, object?>> GetAllRows()
    {
        var rows = new List<Dictionary<string, object?>>();
        for (var i = 1; i <= RowCount; i++)
            rows.Add(GetRow(i));

        return rows;
    }
}
